### PR TITLE
Implement a greedy version of Kernel Inducing Points

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds coresubset algorithm implementing a greedy version of the KIP algorithm from https://arxiv.org/pdf/2011.00050.
Refactors `inverse.py` to be `least_squares.py` to be more generally useful.

### How Has This Been Tested?

Adds corresponding unit tests.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/solvers/coresubset.py [hunk 1]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    ExplicitSizeSolver,
+):
+    r"""
+    Apply GreedyKernelInducingPoints to a supervised dataset.
+
+    GreedyKernelInducingPoints is a deterministic, iterative and greedy approach to
+    build a coreset.
+
+    Given one has an original dataset :math:`\mathcal{D}^{(1)} = \{(x_i, y_i)\}_{i=1}^n`
+    of ``n`` pairs with :math:`x\in\mathbb{R}^d` and :math:`y\in\mathbb{R}^p`, and one
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:543]: should be ``...of :math:`n` pairs...``

File: coreax/solvers/coresubset.py [hunk 2]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    dataset, GreedyKernelInducingPoints selects the next point to minimise the loss
+
+    .. math::
+
+        L(\mathcal{D}^{(1)}, \mathcal{D}^{(2)}) = ||y^{(1)} -
+        K^{(12)}(K^{(22)} + \lambda I_m)^{-1}y^{(2)} ||^2_{\mathbb{R}^n}
+
+    where :math:`y^{(1)}\in\mathbb{R}^n` is the vector of responses from
+    :math:`\mathcal{D}^{(1)}`, :math:`y^{(2)}\in\mathbb{R}^n` is the vector of responses
+    from :math:`\mathcal{D}^{(2)}`,  :math:`K^{(12)} \in \mathbb{R}^{n\times m}`is the
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:555]:
```suggestion
    from :math:`\mathcal{D}^{(2)}`,  :math:`K^{(12)} \in \mathbb{R}^{n\times m}` is the
```
a space is needed for correct formatting

File: coreax/solvers/coresubset.py [hunk 3]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    cross-matrix of kernel evaluations between :math:`\mathcal{D}^{(1)}` and
+    :math:`\mathcal{D}^{(2)}`, :math:`K^{(22)} \in \mathbb{R}^{m\times m}` is the
+    kernel matrix on :math:`\mathcal{D}^{(2)}`,
+    :math:`\lambda I_m \in \mathbb{R}^{m \times m}`is the identity matrix and
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:559]:
```suggestion
    :math:`\lambda I_m \in \mathbb{R}^{m \times m}` is the identity matrix and
```
space needed here also

File: coreax/solvers/coresubset.py [hunk 4]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    :math:`\lambda \in \mathbb{R}_{>0}` is a regularisation parameter.
+
+    The search is performed over the entire dataset.
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:562]: as I understand it, this is only true if `batch_size=None`?

File: coreax/solvers/coresubset.py [hunk 5]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    return loss
+
+
+# pylint: disable=too-many-locals
+class GreedyKernelInducingPoints(
+    RefinementSolver[_SupervisedData, GreedyKernelInducingPointsState],
+    ExplicitSizeSolver,
+):
+    r"""
+    Apply GreedyKernelInducingPoints to a supervised dataset.
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:537]: "GreedyKernelInducingPoints" should either be in `backticks` or have spaces ("Greedy Kernel Inducing Points"). Same comment for the other instances of it in this docstring.

File: tests/unit/test_solvers.py [hunk 6]
```diff
@@ -428,6 +438,146 @@ def solver_factory(self) -> jtu.Partial:
         return jtu.Partial(SteinThinning, coreset_size=coreset_size, kernel=kernel)
 
 
+class TestGreedyKernelInducingPoints(RefinementSolverTest, ExplicitSizeSolverTest):
```
Comment (rg936672) [inline tests/unit/test_solvers.py:441]: Should override `check_solution_invariants` to also check that if `unique=True`, the points are actually unique.

File: coreax/solvers/coresubset.py [hunk 7]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    responses: Array,
+    feature_gramian: Array,
+    regularisation_parameter: float,
+    identity: Array,
+    least_squares_solver: RegularisedLeastSquaresSolver,
+) -> Array:
+    """
+    Given an array of candidate coreset indices, compute the greedy KIP loss for each.
+
+    Primarily intended for use with :class`GreedyKernelInducingPoints.
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:495]:
```suggestion
    Primarily intended for use with :class`GreedyKernelInducingPoints`.
```

**Review by rg936672 — CHANGES_REQUESTED** [review]

File: coreax/solvers/coresubset.py [hunk 8]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    This class works with all children of :class:`~coreax.kernel.Kernel`. Note that
+    GreedyKernelInducingPoints does not support non-uniform weights and will only return
+    coresubsets with uniform weights.
+
+    :param random_key: Key for random number generation
+    :param feature_kernel: :class:`~coreax.kernel.Kernel` instance implementing a kernel
+        function :math:`k: \mathbb{R}^d \times \mathbb{R}^d \rightarrow \mathbb{R}`
+        on the feature space
+    :param regularisation_parameter: Regularisation parameter for stable inversion of
+        feature gram matrix
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:573]: capitalise "Gram"?

File: coreax/solvers/coresubset.py [hunk 9]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    :param least_squares_solver: Instance of
+        :class:`coreax.least_squares.RegularisedLeastSquaresSolver`, default value of
+        :data:`None` uses :class:`coreax.least_squares.MinimalEuclideanNormSolver`
+    """
+
+    random_key: KeyArrayLike
+    feature_kernel: Kernel
+    regularisation_parameter: float = 1e-6
+    unique: bool = True
+    batch_size: Union[int, None] = None
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:587]: Changed it for this line. In #646 see comment: 

"I do think the `Optional` use in the `block_size` and `unroll` type hints needs changing though. For example, with `block_size: Union[int, None, tuple[Optional[int], Optional[int]]]`, if we pass a tuple, it isn't optional to pass an `int`. The tuple should be of type `int` or `None`. It is, however, optional to pass an integer or a tuple, so I think the correct hint is `block_size: Optional[int, tuple[Union[int, None], Union[int, None]]] = None`.

File: coreax/solvers/coresubset.py [hunk 10]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+                (
+                    coresubset.unweighted_indices,
+                    -1 * jnp.ones(pad_size, dtype=jnp.int32),
+                )
+            )
+        elif self.coreset_size < len(coresubset):
+            warn(
+                "Requested coreset size is smaller than input 'coresubset', clipping"
+                + " to the correct size and proceeding...",
+                Warning,
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:637]: Changed this to a `SizeWarning`

File: coreax/solvers/coresubset.py [hunk 11]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    GreedyKernelInducingPoints does not support non-uniform weights and will only return
+    coresubsets with uniform weights.
+
+    :param random_key: Key for random number generation
+    :param feature_kernel: :class:`~coreax.kernel.Kernel` instance implementing a kernel
+        function :math:`k: \mathbb{R}^d \times \mathbb{R}^d \rightarrow \mathbb{R}`
+        on the feature space
+    :param regularisation_parameter: Regularisation parameter for stable inversion of
+        feature gram matrix
+    :param unique: Boolean that enforces the resulting coreset will only contain
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:574]: Added

File: coreax/solvers/coresubset.py [hunk 12]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+
+    Primarily intended for use with :class`GreedyKernelInducingPoints.
+
+    :param candidate_coresets: Array of indices representing all possible "next"
+        coresets
+    :param responses: Array of responses
+    :param feature_gramian: Feature kernel gramian
+    :param regularisation_parameter: Regularisation parameter for stable inversion of
+        array, negative values will be converted to positive
+    :param identity: identity matrix
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:503]: Made this clearer

File: coreax/solvers/coresubset.py (hunk 4 above)
Comment (db091756) [reply coreax/solvers/coresubset.py:562]: Yep, forgot to remove this.

File: tests/unit/test_solvers.py (hunk 6 above)
Comment (db091756) [reply tests/unit/test_solvers.py:441]: Done

File: coreax/solvers/coresubset.py [hunk 13]
```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    ) -> tuple[Coresubset[_SupervisedData], GreedyKernelInducingPointsState]:
+        """
+        Refine a coresubset with 'GreedyKernelInducingPointsState'.
+
+        We first compute the various factors if they are not given in the
+        'solver_state', and then iteratively swap points with the initial coreset,
+        selecting points which minimise the loss.
+
+        :param coresubset: The coresubset to refine
+        :param solver_state: Solution state information, primarily used to cache
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:612]: One big benefit of this structure is that we can easily chain `reduce` and `refine` calls without having to re-initialise the solver class e.g.

`
coreset, state = solver.reduce(dataset)
coreset, state = solver.refine(dataset)
coreset, state = solver.refine(dataset)
`

File: coreax/solvers/coresubset.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rg936672) [reply coreax/solvers/coresubset.py]: That comment is incorrect - `Union[T, None]` is exactly identical to `Optional[T]` ([Python docs](https://docs.python.org/3/library/typing.html#typing.Optional)).

(So, in particular, `Union[int, None, tuple[Optional[int], Optional[int]]]` is the same as `Optional[int, tuple[Optional[int], Optional[int]]]` is the same as `Optional[int, tuple[Union[int, None], Union[int, None]]]`!)

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py]: Updated this now

File: coreax/solvers/coresubset.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (rg936672) [reply coreax/solvers/coresubset.py:620]: Personally, I don't really see the benefit there - it doesn't look like any significant computation is done on initialisation (unless I'm missing something), so there's doesn't appear to be a benefit to retaining the same solver instance and using different cached solver states as opposed to just having multiple solver instances with their own cached states.

File: coreax/solvers/coresubset.py [hunk 14]
```diff
@@ -458,3 +483,274 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+        dataset: _SupervisedData,
+        solver_state: Union[GreedyKernelInducingPointsState, None] = None,
+    ) -> tuple[Coresubset[_SupervisedData], GreedyKernelInducingPointsState]:
+        initial_coresubset = _initial_coresubset(-1, self.coreset_size, dataset)
+        return self.refine(initial_coresubset, solver_state)
+
+    def refine(  # noqa: PLR0915
+        self,
+        coresubset: Coresubset[_SupervisedData],
+        solver_state: Union[GreedyKernelInducingPointsState, None] = None,
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:610]:
```suggestion
        solver_state: Optional[GreedyKernelInducingPointsState] = None,
```

File: coreax/solvers/coresubset.py [hunk 15]
```diff
@@ -458,3 +483,274 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+    regularisation_parameter: float = 1e-6
+    unique: bool = True
+    batch_size: Optional[int] = None
+    least_squares_solver: Optional[RegularisedLeastSquaresSolver] = None
+
+    @override
+    def reduce(
+        self,
+        dataset: _SupervisedData,
+        solver_state: Union[GreedyKernelInducingPointsState, None] = None,
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:602]:
```suggestion
        solver_state: Optional[GreedyKernelInducingPointsState] = None,
```

File: coreax/solvers/coresubset.py [hunk 16]
```diff
@@ -458,3 +483,274 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+        initial_coresubset = _initial_coresubset(-1, self.coreset_size, dataset)
+        return self.refine(initial_coresubset, solver_state)
+
+    def refine(  # noqa: PLR0915
+        self,
+        coresubset: Coresubset[_SupervisedData],
+        solver_state: Union[GreedyKernelInducingPointsState, None] = None,
+    ) -> tuple[Coresubset[_SupervisedData], GreedyKernelInducingPointsState]:
+        """
+        Refine a coresubset with 'GreedyKernelInducingPointsState'.
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:613]:
```suggestion
        Refine a coresubset with `GreedyKernelInducingPoints`.
```

File: coreax/solvers/coresubset.py [hunk 17]
```diff
@@ -458,3 +483,274 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
...
+
+        We first compute the various factors if they are not given in the
+        'solver_state', and then iteratively swap points with the initial coreset,
```
Comment (rg936672) [inline coreax/solvers/coresubset.py:616]:
```suggestion
        `solver_state`, and then iteratively swap points with the initial coreset,
```

File: coreax/solvers/coresubset.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:620]: Note that this is a common pattern for the rest of the solvers in Coreax and was introduced with the refactor

[the remark this answers is not in the document]
Comment (db091756) [reply coreax/solvers/coresubset.py:620]: @tc85324 any other reasons for this I am missing?

[the remark this answers is not in the document]
Comment (rg936672) [reply coreax/solvers/coresubset.py:620]: Ah, I hadn't noticed this was already the pattern used for the other solvers. In that case, leave it as is.

**Review by rg936672 — APPROVED** [review]

**Pull request merged**